File: sources/Tire.cpp

#include "../includes/Tire.hpp"

Tire::Tire(TireType type, Weather weather) : _Type(type), _TurnsUsed(0), _Condition(ConditionStage::Green), _CurrentWeather(weather) {}

void Tire::useTurn()
{
    _TurnsUsed++;
    int green, yellow;
    switch (_Type)
    {
    case TireType::Normal:
        green = 10;
        yellow = green + 25;
        break;
    case TireType::Sprint:
        green = 10;
        yellow = green + 30;
        break;
    case TireType::Wet:
        green = 9;
        yellow = green + 15;
        break;
    }
    if (_TurnsUsed <= green)
        _Condition = ConditionStage::Green;
    if (_TurnsUsed <= yellow)
        _Condition = ConditionStage::Yellow;
    else 
        _Condition = ConditionStage::Red;
}
ConditionStage Tire::getCondition() const { return _Condition; }
int Tire::getMaxWarnings() const
{
    int base = 3;
    if (_Type == TireType::Wet && _CurrentWeather == Weather::Dry)
        base -= 1;
    else if ((_Type == TireType::Normal || _Type == TireType::Sprint) && _CurrentWeather == Weather::Wet)
        base -= 1;
    if (_Condition == ConditionStage::Yellow)
        base -= 1;
    else if (_Condition == ConditionStage::Red)
        base -= 2;
    return (std::max(base, 0));
}
int Tire::getAvailableBrakeDice() const
{
    if (_CurrentWeather == Weather::Wet)
    {
        if (_Type == TireType::Wet)
            return 2;
        else
            return 1;
    }
    return 3;
}
int Tire::getAvailableCoastDice() const
{
    if (_CurrentWeather == Weather::Dry && _Type == TireType::Wet)
        return 1;
    return 2;
}
void Tire::setWeather(Weather w) { _CurrentWeather = w; }
std::string Tire::getTypeAsString() const
{
    switch (_Type)
    {
        case TireType::Normal: return "Normal";
        case TireType::Sprint: return "Sprint";
        case TireType::Wet: return "Wet";
    }
    return "Unknown";
}

File: sources/Tire.ts

export type TireType = "Normal" | "Sprint" | "Wet";
export type Weather = "Dry" | "Wet";
export type ConditionStage = "Green" | "Yellow" | "Red";

export class Tire {
  private type: TireType;
  private turnsUsed = 0;
  private condition: ConditionStage = "Green";
  private currentWeather: Weather;

  constructor(type: TireType, weather: Weather) {
    this.type = type;
    this.currentWeather = weather;
  }

  useTurn(): void {
    this.turnsUsed++;
    let green: number, yellow: number;
    switch (this.type) {
      case "Normal": green = 10; yellow = 35; break;
      case "Sprint": green = 10; yellow = 40; break;
      case "Wet": green = 9; yellow = 24; break;
    }

    if (this.turnsUsed <= green)
      this.condition = "Green";
    else if (this.turnsUsed <= yellow)
      this.condition = "Yellow";
    else
      this.condition = "Red";
  }

  getCondition(): ConditionStage {
    return this.condition;
  }

  getMaxWarnings(): number {
    let base = 3;
    if (this.type === "Wet" && this.currentWeather === "Dry")
      base--;
    else if ((this.type === "Normal" || this.type === "Sprint") && this.currentWeather === "Wet")
      base--;

    if (this.condition === "Yellow")
      base--;
    else if (this.condition === "Red")
      base -= 2;

    return Math.max(base, 0);
  }

  getAvailableBrakeDice(): number {
    if (this.currentWeather === "Wet") {
      return this.type === "Wet" ? 2 : 1;
    }
    return 3;
  }

  getAvailableCoastDice(): number {
    if (this.currentWeather === "Dry" && this.type === "Wet")
      return 1;
    return 2;
  }

  setWeather(w: Weather): void {
    this.currentWeather = w;
  }

  getTypeAsString(): string {
    return this.type;
  }
}
